Move iiwa and WSG lcm handlers

sammy-tri opened the pull request:
Split the LCM-related stuff for the iiwa and the WSG into their own files instead of living with the rest of the simulation.

This is in support of an upcoming PR to create a combined LCM controlled iiwa+WSG simulation.

<!-- Reviewable:start -->
---
This change is [<img src="https://reviewable.io/review_button.svg" height="34" align="absmiddle" alt="Reviewable"/>](https://reviewable.io/reviews/robotlocomotion/drake/4405)
<!-- Reviewable:end -->

sammy-tri commented:
+@liangfok for feature review

---

Review status: 0 of 8 files reviewed at latest revision, all discussions resolved.

---



*Comments from [Reviewable](https://reviewable.io:443/reviews/robotlocomotion/drake/4405#-:-KYP3QJz36S0iRudTpih:b-ui3opy)*
<!-- Sent from Reviewable.io -->

sammy-tri commented:
+@ggould-tri for platform review (looks like he's taking Thursday this week)

---

Review status: 0 of 8 files reviewed at latest revision, all discussions resolved.

---



*Comments from [Reviewable](https://reviewable.io:443/reviews/robotlocomotion/drake/4405#-:-KYP3aUTAOk6WdC9qi0c:b-4metbx)*
<!-- Sent from Reviewable.io -->

liangfok commented:
@sammy-tri, I'm checking in a partial review. and will return to this after addressing some other high priority tasks. Hopefully some of my initial comments can be addressed in the mean time.

---

Reviewed 5 of 8 files at r1.
Review status: 5 of 8 files reviewed at latest revision, 8 unresolved discussions.

---

*[drake/examples/kuka_iiwa_arm/iiwa_lcm.h, line 17 at r1](https://reviewable.io:443/reviews/robotlocomotion/drake/4405#-KYP5kQS40yS9yU_6hkV:-KYP5kQS3Iup33MynhrR:b-na99nt) ([raw file](https://github.com/robotlocomotion/drake/blob/7b3e84d1a61022dd2808fdc7a16847d0258d32cb/drake/examples/kuka_iiwa_arm/iiwa_lcm.h#L17)):*
> ```objc
> class IiwaCommandReceiver : public systems::LeafSystem<double> {
>  public:
>   explicit IiwaCommandReceiver(int num_joints);
> ```

BTW, since this is an `IiwaCommandReceiver`, it is clearly specific to the KUKA iiwa robot. Thus, does it make sense to take the number of joints as a parameter? Is the intent that this class eventually be generalized to support other robots? Or, is it expected that the KUKA iiwa robot will have a variable number of joints?

(Same comment about the Schunk WSG LCM class)

---

*[drake/examples/kuka_iiwa_arm/iiwa_lcm.h, line 19 at r1](https://reviewable.io:443/reviews/robotlocomotion/drake/4405#-KYP6FvQ9DQuGv3zDTek:-KYP6FvQDDBy1eeK0UMD:b5y62at) ([raw file](https://github.com/robotlocomotion/drake/blob/7b3e84d1a61022dd2808fdc7a16847d0258d32cb/drake/examples/kuka_iiwa_arm/iiwa_lcm.h#L19)):*
> ```objc
>   explicit IiwaCommandReceiver(int num_joints);
> 
>   void EvalOutput(const systems::Context<double>& context,
> ```

Should document that the output vector will be all zero while awaiting the first `lcmt_iiwa_command` message to arrive.

---

*[drake/examples/kuka_iiwa_arm/iiwa_lcm.h, line 25 at r1](https://reviewable.io:443/reviews/robotlocomotion/drake/4405#-KYP5FxaBitCh1jZiVG3:-KYP5Fxa-qMI2arVq7nO:b-p0ot0v) ([raw file](https://github.com/robotlocomotion/drake/blob/7b3e84d1a61022dd2808fdc7a16847d0258d32cb/drake/examples/kuka_iiwa_arm/iiwa_lcm.h#L25)):*
> ```objc
> /// Sends lcmt_iiwa_status messages.  This system has two input ports,
> /// one for the current state of the plant and one for the most
> /// recently received command.
> ```

BTW, consider adding a sentence here saying the size of the state input port must be twice the size of the command input port.

(Also, see comment above about whether the number of joints should be hard-coded since this class is specific to the KUKA iiwa robot arm.)

---

*[drake/examples/schunk_wsg/schunk_wsg_lcm.cc, line 28 at r1](https://reviewable.io:443/reviews/robotlocomotion/drake/4405#-KYPS8QT9xIgUskwHq4K:-KYPS8QT-lVwN_g5oCEo:bx4c1xt) ([raw file](https://github.com/robotlocomotion/drake/blob/7b3e84d1a61022dd2808fdc7a16847d0258d32cb/drake/examples/schunk_wsg/schunk_wsg_lcm.cc#L28)):*
> ```cpp
>   this->DeclareAbstractInputPort();
>   this->DeclareInputPort(
>       systems::kVectorValued, input_size);
> ```

BTW, can this line be combined with the preceding line?

---

*[drake/examples/schunk_wsg/schunk_wsg_lcm.cc, line 40 at r1](https://reviewable.io:443/reviews/robotlocomotion/drake/4405#-KYPTJWt7PVJ1IlCT5SG:-KYPTJWt87w38HnArngZ:b-z80wza) ([raw file](https://github.com/robotlocomotion/drake/blob/7b3e84d1a61022dd2808fdc7a16847d0258d32cb/drake/examples/schunk_wsg/schunk_wsg_lcm.cc#L40)):*
> ```cpp
>   const systems::BasicVector<double>* state =
>       this->EvalVectorInput(context, 1);
>   const double cur_position = state->GetAtIndex(position_index_);
> ```

BTW, I find it odd for the context of this system to contain more state than what this system cares about. Wouldn't a better design be to require that the context for this system contain just the WSG gripper's state and then to use a `Demultiplexer` to extract out the relevant signal for this system?

---

*[drake/examples/schunk_wsg/schunk_wsg_lcm.h, line 24 at r1](https://reviewable.io:443/reviews/robotlocomotion/drake/4405#-KYPSY3rDGio1lvQNK7U:-KYPSY3rBZKdMolTv0YE:b-tj3f4l) ([raw file](https://github.com/robotlocomotion/drake/blob/7b3e84d1a61022dd2808fdc7a16847d0258d32cb/drake/examples/schunk_wsg/schunk_wsg_lcm.h#L24)):*
> ```objc
> class SchunkWsgTrajectoryGenerator : public systems::LeafSystem<double> {
>  public:
>   SchunkWsgTrajectoryGenerator(int input_size, int position_index);
> ```

Please document the meaning of parameter `position_index`.

Same comment here regarding whether parameter `input_size` is necessary given that this class is specific to the Schunk WSG gripper.

---

*[drake/examples/schunk_wsg/schunk_wsg_lcm.h, line 44 at r1](https://reviewable.io:443/reviews/robotlocomotion/drake/4405#-KYPSl0P8PPUw-plz7fw:-KYPSl0PAOgbqZUu5oRw:b2srgxi) ([raw file](https://github.com/robotlocomotion/drake/blob/7b3e84d1a61022dd2808fdc7a16847d0258d32cb/drake/examples/schunk_wsg/schunk_wsg_lcm.h#L44)):*
> ```objc
> 
>   std::unique_ptr<systems::DifferenceState<double>>
>   AllocateDifferenceState() const override;
> ```

Line continuations should be indented by four spaces.

---

*[drake/examples/schunk_wsg/schunk_wsg_lcm.h, line 63 at r1](https://reviewable.io:443/reviews/robotlocomotion/drake/4405#-KYPT1EL8s6YQ2Ah7vz0:-KYPT1EL13-MKLHgUkQV:b-piuv2v) ([raw file](https://github.com/robotlocomotion/drake/blob/7b3e84d1a61022dd2808fdc7a16847d0258d32cb/drake/examples/schunk_wsg/schunk_wsg_lcm.h#L63)):*
> ```objc
> };
> 
> /// Sends status messages for a Schunk WSG.  This system has one input
> ```

BTW, consider replacing "status" with `lcmt_schunk_wsg_status` to be more specific.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/robotlocomotion/drake/4405#-:-KYPUDv66ylB0QLqpQSv:b4g415p)*
<!-- Sent from Reviewable.io -->

sammy-tri commented:
Review status: 3 of 8 files reviewed at latest revision, 8 unresolved discussions.

---

*[drake/examples/kuka_iiwa_arm/iiwa_lcm.h, line 17 at r1](https://reviewable.io:443/reviews/robotlocomotion/drake/4405#-KYP5kQS40yS9yU_6hkV:-KYPVl0A8P49omkNRJ2A:bz4qteo) ([raw file](https://github.com/robotlocomotion/drake/blob/7b3e84d1a61022dd2808fdc7a16847d0258d32cb/drake/examples/kuka_iiwa_arm/iiwa_lcm.h#L17)):*
<details><summary><i>Previously, liangfok (Chien-Liang Fok) wrote…</i></summary><blockquote>

BTW, since this is an `IiwaCommandReceiver`, it is clearly specific to the KUKA iiwa robot. Thus, does it make sense to take the number of joints as a parameter? Is the intent that this class eventually be generalized to support other robots? Or, is it expected that the KUKA iiwa robot will have a variable number of joints?

(Same comment about the Schunk WSG LCM class)
</blockquote></details>

I was concerned that hardcoding 7 or kNumJoints would be frowned upon...  but I can't think of a reason why it would ever change...

---

*[drake/examples/kuka_iiwa_arm/iiwa_lcm.h, line 19 at r1](https://reviewable.io:443/reviews/robotlocomotion/drake/4405#-KYP6FvQ9DQuGv3zDTek:-KYPW47h9G50v54Kkqhm:b28aa83) ([raw file](https://github.com/robotlocomotion/drake/blob/7b3e84d1a61022dd2808fdc7a16847d0258d32cb/drake/examples/kuka_iiwa_arm/iiwa_lcm.h#L19)):*
<details><summary><i>Previously, liangfok (Chien-Liang Fok) wrote…</i></summary><blockquote>

Should document that the output vector will be all zero while awaiting the first `lcmt_iiwa_command` message to arrive.
</blockquote></details>

Done.  (in class comment about the port)

---

*[drake/examples/kuka_iiwa_arm/iiwa_lcm.h, line 25 at r1](https://reviewable.io:443/reviews/robotlocomotion/drake/4405#-KYP5FxaBitCh1jZiVG3:-KYPWTe2Bwy9W97L_E_K:b-896fix) ([raw file](https://github.com/robotlocomotion/drake/blob/7b3e84d1a61022dd2808fdc7a16847d0258d32cb/drake/examples/kuka_iiwa_arm/iiwa_lcm.h#L25)):*
<details><summary><i>Previously, liangfok (Chien-Liang Fok) wrote…</i></summary><blockquote>

BTW, consider adding a sentence here saying the size of the state input port must be twice the size of the command input port.

(Also, see comment above about whether the number of joints should be hard-coded since this class is specific to the KUKA iiwa robot arm.)
</blockquote></details>

Done.

---

*[drake/examples/schunk_wsg/schunk_wsg_lcm.cc, line 28 at r1](https://reviewable.io:443/reviews/robotlocomotion/drake/4405#-KYPS8QT9xIgUskwHq4K:-KYPWYuUCjhUKfeS5t-H:b-896fix) ([raw file](https://github.com/robotlocomotion/drake/blob/7b3e84d1a61022dd2808fdc7a16847d0258d32cb/drake/examples/schunk_wsg/schunk_wsg_lcm.cc#L28)):*
<details><summary><i>Previously, liangfok (Chien-Liang Fok) wrote…</i></summary><blockquote>

BTW, can this line be combined with the preceding line?
</blockquote></details>

Done.

---

*[drake/examples/schunk_wsg/schunk_wsg_lcm.cc, line 40 at r1](https://reviewable.io:443/reviews/robotlocomotion/drake/4405#-KYPTJWt7PVJ1IlCT5SG:-KYPWduM0cTgN3VGPsRb:b-cbbdmd) ([raw file](https://github.com/robotlocomotion/drake/blob/7b3e84d1a61022dd2808fdc7a16847d0258d32cb/drake/examples/schunk_wsg/schunk_wsg_lcm.cc#L40)):*
<details><summary><i>Previously, liangfok (Chien-Liang Fok) wrote…</i></summary><blockquote>

BTW, I find it odd for the context of this system to contain more state than what this system cares about. Wouldn't a better design be to require that the context for this system contain just the WSG gripper's state and then to use a `Demultiplexer` to extract out the relevant signal for this system?
</blockquote></details>

The WSG model has only one actuator (the left finger which connects to the right through a loop joint, IIRC), but several positions, most of which we don't care about.  I didn't think of a Demultiplexer for the WSG-only case, though a future PR for the iiwa+WSG system does basically do this.

---

*[drake/examples/schunk_wsg/schunk_wsg_lcm.h, line 24 at r1](https://reviewable.io:443/reviews/robotlocomotion/drake/4405#-KYPSY3rDGio1lvQNK7U:-KYPXY9y0C2EpHf8XFam:b-zayx0u) ([raw file](https://github.com/robotlocomotion/drake/blob/7b3e84d1a61022dd2808fdc7a16847d0258d32cb/drake/examples/schunk_wsg/schunk_wsg_lcm.h#L24)):*
<details><summary><i>Previously, liangfok (Chien-Liang Fok) wrote…</i></summary><blockquote>

Please document the meaning of parameter `position_index`.

Same comment here regarding whether parameter `input_size` is necessary given that this class is specific to the Schunk WSG gripper.
</blockquote></details>

input_size is more meaningful here in that it's intended to handle either the entire state of a WSG, or possibly a demultiplexed value with only one entry in the input state.  Comment added.

---

*[drake/examples/schunk_wsg/schunk_wsg_lcm.h, line 44 at r1](https://reviewable.io:443/reviews/robotlocomotion/drake/4405#-KYPSl0P8PPUw-plz7fw:-KYPXyKe9volq3aJZZYO:b21knb0) ([raw file](https://github.com/robotlocomotion/drake/blob/7b3e84d1a61022dd2808fdc7a16847d0258d32cb/drake/examples/schunk_wsg/schunk_wsg_lcm.h#L44)):*
<details><summary><i>Previously, liangfok (Chien-Liang Fok) wrote…</i></summary><blockquote>

Line continuations should be indented by four spaces.
</blockquote></details>

Emacs disagrees.  Switched to the (indented) clang-format style.

---

*[drake/examples/schunk_wsg/schunk_wsg_lcm.h, line 63 at r1](https://reviewable.io:443/reviews/robotlocomotion/drake/4405#-KYPT1EL8s6YQ2Ah7vz0:-KYPY6tzC9jz5_M0LbBE:b-896fix) ([raw file](https://github.com/robotlocomotion/drake/blob/7b3e84d1a61022dd2808fdc7a16847d0258d32cb/drake/examples/schunk_wsg/schunk_wsg_lcm.h#L63)):*
<details><summary><i>Previously, liangfok (Chien-Liang Fok) wrote…</i></summary><blockquote>

BTW, consider replacing "status" with `lcmt_schunk_wsg_status` to be more specific.
</blockquote></details>

Done.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/robotlocomotion/drake/4405)*
<!-- Sent from Reviewable.io -->

liangfok commented:
@sammy-tri, I completed my first round feature review. PTAL. 

---

Reviewed 2 of 8 files at r1, 3 of 3 files at r2.
Review status: all files reviewed at latest revision, 12 unresolved discussions.

---

*[drake/examples/kuka_iiwa_arm/iiwa_lcm.cc, line 15 at r2](https://reviewable.io:443/reviews/robotlocomotion/drake/4405#-KYPg5F_6-0JGttk3UcR:-KYPg5F_EmDWS8IgqToM:busr04o) ([raw file](https://github.com/robotlocomotion/drake/blob/d83d147a017e12987f450795da93abbd9977b372/drake/examples/kuka_iiwa_arm/iiwa_lcm.cc#L15)):*
> ```cpp
> IiwaCommandReceiver::IiwaCommandReceiver(int num_joints) {
>   this->DeclareAbstractInputPort();
>   this->DeclareOutputPort(systems::kVectorValued, num_joints);
> ```

BTW, this returns a const reference to a `SystemPortDescriptor`, which contains the actual port number. Consider saving this port number as a class member variable, and using it when calling `GetMutableOutputVector()` below. This would be less brittle than the assumption that the port number is zero, which is what's currently being done.

(Same command regarding the call to `DeclareAbstractInputPort()` above.)

---

*[drake/examples/kuka_iiwa_arm/iiwa_lcm.cc, line 26 at r2](https://reviewable.io:443/reviews/robotlocomotion/drake/4405#-KYP_PtF3B04HKK9ub_J:-KYP_PtF8uy5z951TQOX:b-5cviev) ([raw file](https://github.com/robotlocomotion/drake/blob/d83d147a017e12987f450795da93abbd9977b372/drake/examples/kuka_iiwa_arm/iiwa_lcm.cc#L26)):*
> ```cpp
> 
>   // If we're using a default constructed message (haven't received
>   // a command yet), just return 0.
> ```

BTW, "return 0." should be "output a vector of zeros."

---

*[drake/examples/kuka_iiwa_arm/iiwa_lcm.cc, line 35 at r2](https://reviewable.io:443/reviews/robotlocomotion/drake/4405#-KYP_Jrb3HtIEMeD-TMJ:-KYP_JrbAvfRvWc1cHxs:b-bltkyz) ([raw file](https://github.com/robotlocomotion/drake/blob/d83d147a017e12987f450795da93abbd9977b372/drake/examples/kuka_iiwa_arm/iiwa_lcm.cc#L35)):*
> ```cpp
>   }
> 
>   // TODO(sam.creasey) Support torque control some day.
> ```

BTW, "some day" should be "someday" based on this website:

http://grammarist.com/usage/someday-some-day/

---

*[drake/examples/kuka_iiwa_arm/iiwa_lcm.cc, line 66 at r2](https://reviewable.io:443/reviews/robotlocomotion/drake/4405#-KYPb88N5VAXPNPTvJcY:-KYPb88NBbmis9YM7uR1:br3b57y) ([raw file](https://github.com/robotlocomotion/drake/blob/d83d147a017e12987f450795da93abbd9977b372/drake/examples/kuka_iiwa_arm/iiwa_lcm.cc#L66)):*
> ```cpp
> void IiwaStatusSender::EvalOutput(
>     const Context<double>& context,
>     SystemOutput<double>* output) const {
> ```

BTW, consider combining this line with the previous line since it will still be < 80 characters wide.

---

*[drake/examples/kuka_iiwa_arm/iiwa_lcm.h, line 17 at r1](https://reviewable.io:443/reviews/robotlocomotion/drake/4405#-KYP5kQS40yS9yU_6hkV:-KYPZXvlBqWSOEFo0HOa:b-7xxfba) ([raw file](https://github.com/robotlocomotion/drake/blob/7b3e84d1a61022dd2808fdc7a16847d0258d32cb/drake/examples/kuka_iiwa_arm/iiwa_lcm.h#L17)):*
<details><summary><i>Previously, sammy-tri (Sam Creasey) wrote…</i></summary><blockquote>

I was concerned that hardcoding 7 or kNumJoints would be frowned upon...  but I can't think of a reason why it would ever change...
</blockquote></details>

Since we can't think of a reason the number of joints would change, I recommend removing this parameter and using a `kNumJoints` constant since that will eliminate one source of user error.

---

*[drake/examples/kuka_iiwa_arm/iiwa_lcm.h, line 16 at r2](https://reviewable.io:443/reviews/robotlocomotion/drake/4405#-KYPZtqN704IwlB4lqn8:-KYPZtqNEFk8y4QaOdfX:b-6s97ab) ([raw file](https://github.com/robotlocomotion/drake/blob/d83d147a017e12987f450795da93abbd9977b372/drake/examples/kuka_iiwa_arm/iiwa_lcm.h#L16)):*
> ```objc
> /// for each joint.  All values on the output port will be zero if no
> /// LCM message has arrived with a command yet.
> class IiwaCommandReceiver : public systems::LeafSystem<double> {
> ```

BTW, for my own educational purposes: Suppose we scale up to have simulations containing multiple KUKA iiwa robots. Will we instantiate multiple `IiwaCommandReceiver` objects, one for each instance?

---

*[drake/examples/schunk_wsg/schunk_wsg_lcm.cc, line 40 at r1](https://reviewable.io:443/reviews/robotlocomotion/drake/4405#-KYPTJWt7PVJ1IlCT5SG:-KYPc6Bm7vp9fjoibIGn:b-stnoss) ([raw file](https://github.com/robotlocomotion/drake/blob/7b3e84d1a61022dd2808fdc7a16847d0258d32cb/drake/examples/schunk_wsg/schunk_wsg_lcm.cc#L40)):*
<details><summary><i>Previously, sammy-tri (Sam Creasey) wrote…</i></summary><blockquote>

The WSG model has only one actuator (the left finger which connects to the right through a loop joint, IIRC), but several positions, most of which we don't care about.  I didn't think of a Demultiplexer for the WSG-only case, though a future PR for the iiwa+WSG system does basically do this.
</blockquote></details>

OK I am hoping that in a future PR, we can eliminate the need for a `position_index_` parameter 

---

*[drake/examples/schunk_wsg/schunk_wsg_lcm.h, line 18 at r2](https://reviewable.io:443/reviews/robotlocomotion/drake/4405#-KYPc__8DIVAznSu5Lq8:-KYPc__81aTwe1eVqu3V:bfz20xf) ([raw file](https://github.com/robotlocomotion/drake/blob/d83d147a017e12987f450795da93abbd9977b372/drake/examples/schunk_wsg/schunk_wsg_lcm.h#L18)):*
> ```objc
> // state of the gripper (particularly the maximum force).
> 
> /// Recieves commands for a Schunk WSG (input port 0) along with the
> ```

Typo: "Recieves"

---

*[drake/examples/schunk_wsg/schunk_wsg_lcm.h, line 24 at r2](https://reviewable.io:443/reviews/robotlocomotion/drake/4405#-KYPcyH54GmILfmG7rAp:-KYPcyH5ECF7nzvEZ2mh:b54xhzh) ([raw file](https://github.com/robotlocomotion/drake/blob/d83d147a017e12987f450795da93abbd9977b372/drake/examples/schunk_wsg/schunk_wsg_lcm.h#L24)):*
> ```objc
> class SchunkWsgTrajectoryGenerator : public systems::LeafSystem<double> {
>  public:
>   /// @param input_size The size of the state input port to create.
> ```

BTW, consider adding a comment here saying the input size can either be the entire state of a Schunk WSG gripper or a single value that is the state of the actuated joint.

---

*[drake/examples/schunk_wsg/schunk_wsg_lcm.h, line 25 at r2](https://reviewable.io:443/reviews/robotlocomotion/drake/4405#-KYPdB1h7ABz5SFEoVsA:-KYPdB1h0LmeeyWcpIFe:bs8svzt) ([raw file](https://github.com/robotlocomotion/drake/blob/d83d147a017e12987f450795da93abbd9977b372/drake/examples/schunk_wsg/schunk_wsg_lcm.h#L25)):*
> ```objc
>  public:
>   /// @param input_size The size of the state input port to create.
>   /// @param position_index The the index in the state input vector
> ```

BTW, consider changing this parameter to be called `actuated_position_index` for extra clarity.

---

*[drake/examples/schunk_wsg/schunk_wsg_lcm.h, line 25 at r2](https://reviewable.io:443/reviews/robotlocomotion/drake/4405#-KYPdRk68IvTmTouWLp_:-KYPdRk65euSaoUjmZ2S:b-p68kl) ([raw file](https://github.com/robotlocomotion/drake/blob/d83d147a017e12987f450795da93abbd9977b372/drake/examples/schunk_wsg/schunk_wsg_lcm.h#L25)):*
> ```objc
>  public:
>   /// @param input_size The size of the state input port to create.
>   /// @param position_index The the index in the state input vector
> ```

Typo: "The the".

---

*[drake/examples/schunk_wsg/schunk_wsg_lcm.h, line 67 at r2](https://reviewable.io:443/reviews/robotlocomotion/drake/4405#-KYPdt6Q5JNxmFA49V1l:-KYPdt6R5JWIZx_yQyVY:b-5eqy03) ([raw file](https://github.com/robotlocomotion/drake/blob/d83d147a017e12987f450795da93abbd9977b372/drake/examples/schunk_wsg/schunk_wsg_lcm.h#L67)):*
> ```objc
> 
> /// Sends lcmt_schunk_wsg_status messages for a Schunk WSG.  This
> /// system has one input port for the current state of the plant.
> ```

I believe the word "plant" is undefined here. Can you elaborate? For example, is it assumed that this system is "wired" up to a `RigidBodyPlant` that contains a model of the Schunk WSG?

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/robotlocomotion/drake/4405#-:-KYPgoyfCAvBDGLQEYME:bp31lrx)*
<!-- Sent from Reviewable.io -->

liangfok commented:
Review status: all files reviewed at latest revision, 12 unresolved discussions.

---

*[a discussion](https://reviewable.io:443/reviews/robotlocomotion/drake/4405#-KYPhoV-2VD62-chCJfU:-KYPhoV-7sRe5vEerR8o:b-78wft7) (no related file):*
BTW, I just noticed that there are no unit tests being added or modified in this PR. Is there sufficient unit test coverage for the code touched by this PR?

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/robotlocomotion/drake/4405)*
<!-- Sent from Reviewable.io -->

sammy-tri commented:
Review status: 5 of 8 files reviewed at latest revision, 11 unresolved discussions.

---

*[a discussion](https://reviewable.io:443/reviews/robotlocomotion/drake/4405#-KYPhoV-2VD62-chCJfU:-KYPyqafEWzT-Cpl-A8X:b-ki99ey) (no related file):*
<details><summary><i>Previously, liangfok (Chien-Liang Fok) wrote…</i></summary><blockquote>

BTW, I just noticed that there are no unit tests being added or modified in this PR. Is there sufficient unit test coverage for the code touched by this PR?
</blockquote></details>

There is definitely not sufficient unit test coverage...  and I'd argue that as robot-specific LCM marshalling code, perhaps it's not needed, but that seems like it might be a weak argument given that the WSG trajectory generator has grown non-trivial functionality.  Will investigate.

---

*[drake/examples/kuka_iiwa_arm/iiwa_lcm.cc, line 15 at r2](https://reviewable.io:443/reviews/robotlocomotion/drake/4405#-KYPg5F_6-0JGttk3UcR:-KYPzJhyDEJy09rmd30p:b-nsx9b7) ([raw file](https://github.com/robotlocomotion/drake/blob/d83d147a017e12987f450795da93abbd9977b372/drake/examples/kuka_iiwa_arm/iiwa_lcm.cc#L15)):*
<details><summary><i>Previously, liangfok (Chien-Liang Fok) wrote…</i></summary><blockquote>

BTW, this returns a const reference to a `SystemPortDescriptor`, which contains the actual port number. Consider saving this port number as a class member variable, and using it when calling `GetMutableOutputVector()` below. This would be less brittle than the assumption that the port number is zero, which is what's currently being done.

(Same command regarding the call to `DeclareAbstractInputPort()` above.)
</blockquote></details>

This seems like a bit overkill for for a LeafSystem which only declares one input and one output port?  Certainly plenty of the implementations in framework/primitives assume zero is safe to use blindly in this case.

---

*[drake/examples/kuka_iiwa_arm/iiwa_lcm.cc, line 26 at r2](https://reviewable.io:443/reviews/robotlocomotion/drake/4405#-KYP_PtF3B04HKK9ub_J:-KYPzczZ-3HHSSg_qPOo:b-896fix) ([raw file](https://github.com/robotlocomotion/drake/blob/d83d147a017e12987f450795da93abbd9977b372/drake/examples/kuka_iiwa_arm/iiwa_lcm.cc#L26)):*
<details><summary><i>Previously, liangfok (Chien-Liang Fok) wrote…</i></summary><blockquote>

BTW, "return 0." should be "output a vector of zeros."
</blockquote></details>

Done.

---

*[drake/examples/kuka_iiwa_arm/iiwa_lcm.cc, line 35 at r2](https://reviewable.io:443/reviews/robotlocomotion/drake/4405#-KYP_Jrb3HtIEMeD-TMJ:-KYPzfoW-GKnPkATqOGE:b-gz5vwd) ([raw file](https://github.com/robotlocomotion/drake/blob/d83d147a017e12987f450795da93abbd9977b372/drake/examples/kuka_iiwa_arm/iiwa_lcm.cc#L35)):*
<details><summary><i>Previously, liangfok (Chien-Liang Fok) wrote…</i></summary><blockquote>

BTW, "some day" should be "someday" based on this website:

http://grammarist.com/usage/someday-some-day/
</blockquote></details>

Done.  (well, removed since it didn't seem to be adding much)

---

*[drake/examples/kuka_iiwa_arm/iiwa_lcm.cc, line 66 at r2](https://reviewable.io:443/reviews/robotlocomotion/drake/4405#-KYPb88N5VAXPNPTvJcY:-KYPzoCX735j4DeRZb1P:b-896fix) ([raw file](https://github.com/robotlocomotion/drake/blob/d83d147a017e12987f450795da93abbd9977b372/drake/examples/kuka_iiwa_arm/iiwa_lcm.cc#L66)):*
<details><summary><i>Previously, liangfok (Chien-Liang Fok) wrote…</i></summary><blockquote>

BTW, consider combining this line with the previous line since it will still be < 80 characters wide.
</blockquote></details>

Done.

---

*[drake/examples/kuka_iiwa_arm/iiwa_lcm.h, line 16 at r2](https://reviewable.io:443/reviews/robotlocomotion/drake/4405#-KYPZtqN704IwlB4lqn8:-KYPzqTv0rQrNGezhAGU:b-tawghq) ([raw file](https://github.com/robotlocomotion/drake/blob/d83d147a017e12987f450795da93abbd9977b372/drake/examples/kuka_iiwa_arm/iiwa_lcm.h#L16)):*
<details><summary><i>Previously, liangfok (Chien-Liang Fok) wrote…</i></summary><blockquote>

BTW, for my own educational purposes: Suppose we scale up to have simulations containing multiple KUKA iiwa robots. Will we instantiate multiple `IiwaCommandReceiver` objects, one for each instance?
</blockquote></details>

That would be my assumption, since it's also what I think we'd do if we had multiple actual iiwa arms...   though given that LCM is typically multicast, running multiple iiwa arms would take some additional code changes or careful separation of the LCM domains (e.g. different UDP port numbers).

---

*[drake/examples/schunk_wsg/schunk_wsg_lcm.h, line 18 at r2](https://reviewable.io:443/reviews/robotlocomotion/drake/4405#-KYPc__8DIVAznSu5Lq8:-KYQ-L26A0L1t_v9jA_0:b-896fix) ([raw file](https://github.com/robotlocomotion/drake/blob/d83d147a017e12987f450795da93abbd9977b372/drake/examples/schunk_wsg/schunk_wsg_lcm.h#L18)):*
<details><summary><i>Previously, liangfok (Chien-Liang Fok) wrote…</i></summary><blockquote>

Typo: "Recieves"
</blockquote></details>

Done.

---

*[drake/examples/schunk_wsg/schunk_wsg_lcm.h, line 24 at r2](https://reviewable.io:443/reviews/robotlocomotion/drake/4405#-KYPcyH54GmILfmG7rAp:-KYQ-r8n5qCR0jGzlir_:bz88tee) ([raw file](https://github.com/robotlocomotion/drake/blob/d83d147a017e12987f450795da93abbd9977b372/drake/examples/schunk_wsg/schunk_wsg_lcm.h#L24)):*
<details><summary><i>Previously, liangfok (Chien-Liang Fok) wrote…</i></summary><blockquote>

BTW, consider adding a comment here saying the input size can either be the entire state of a Schunk WSG gripper or a single value that is the state of the actuated joint.
</blockquote></details>

Added some extra language to that effect

---

*[drake/examples/schunk_wsg/schunk_wsg_lcm.h, line 25 at r2](https://reviewable.io:443/reviews/robotlocomotion/drake/4405#-KYPdRk68IvTmTouWLp_:-KYQ-xwXDGdiNDYZi8xT:b-896fix) ([raw file](https://github.com/robotlocomotion/drake/blob/d83d147a017e12987f450795da93abbd9977b372/drake/examples/schunk_wsg/schunk_wsg_lcm.h#L25)):*
<details><summary><i>Previously, liangfok (Chien-Liang Fok) wrote…</i></summary><blockquote>

Typo: "The the".
</blockquote></details>

Done.

---

*[drake/examples/schunk_wsg/schunk_wsg_lcm.h, line 67 at r2](https://reviewable.io:443/reviews/robotlocomotion/drake/4405#-KYPdt6Q5JNxmFA49V1l:-KYQ0DTq5pyFO2VupXme:b-oc87yb) ([raw file](https://github.com/robotlocomotion/drake/blob/d83d147a017e12987f450795da93abbd9977b372/drake/examples/schunk_wsg/schunk_wsg_lcm.h#L67)):*
<details><summary><i>Previously, liangfok (Chien-Liang Fok) wrote…</i></summary><blockquote>

I believe the word "plant" is undefined here. Can you elaborate? For example, is it assumed that this system is "wired" up to a `RigidBodyPlant` that contains a model of the Schunk WSG?
</blockquote></details>

Stated assumption in comment.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/robotlocomotion/drake/4405)*
<!-- Sent from Reviewable.io -->

liangfok commented:
Reviewed 2 of 3 files at r3.
Review status: 7 of 8 files reviewed at latest revision, 4 unresolved discussions.

---

*[a discussion](https://reviewable.io:443/reviews/robotlocomotion/drake/4405#-KYPhoV-2VD62-chCJfU:-KYQ3dm02qVf_jjM9Wio:bbgrkzq) (no related file):*
<details><summary><i>Previously, sammy-tri (Sam Creasey) wrote…</i></summary><blockquote>

There is definitely not sufficient unit test coverage...  and I'd argue that as robot-specific LCM marshalling code, perhaps it's not needed, but that seems like it might be a weak argument given that the WSG trajectory generator has grown non-trivial functionality.  Will investigate.
</blockquote></details>

OK thanks.

---

*[drake/examples/kuka_iiwa_arm/iiwa_lcm.cc, line 15 at r2](https://reviewable.io:443/reviews/robotlocomotion/drake/4405#-KYPg5F_6-0JGttk3UcR:-KYQ1qdP6t6nNfssczTJ:b-m6lm5) ([raw file](https://github.com/robotlocomotion/drake/blob/d83d147a017e12987f450795da93abbd9977b372/drake/examples/kuka_iiwa_arm/iiwa_lcm.cc#L15)):*
<details><summary><i>Previously, sammy-tri (Sam Creasey) wrote…</i></summary><blockquote>

This seems like a bit overkill for for a LeafSystem which only declares one input and one output port?  Certainly plenty of the implementations in framework/primitives assume zero is safe to use blindly in this case.
</blockquote></details>

OK. Let's keep this in mind as a possible future enhancement. I feel our current usage of port number makes potentially brittle assumptions about how `systems::System` implements ports. As it stands, `systems::Systems::DeclareOutputPort()` does not promise that the first declared port will be at index 0!

---

*[drake/examples/kuka_iiwa_arm/iiwa_lcm.cc, line 26 at r2](https://reviewable.io:443/reviews/robotlocomotion/drake/4405#-KYP_PtF3B04HKK9ub_J:-KYQ2U5e-9DZ-RojG0ax:b-vsmnq1) ([raw file](https://github.com/robotlocomotion/drake/blob/d83d147a017e12987f450795da93abbd9977b372/drake/examples/kuka_iiwa_arm/iiwa_lcm.cc#L26)):*
<details><summary><i>Previously, sammy-tri (Sam Creasey) wrote…</i></summary><blockquote>

Done.
</blockquote></details>

BTW: nit: this method actually returns `void`... it _outputs a vector_ of zeros...

---

*[drake/examples/kuka_iiwa_arm/iiwa_lcm.h, line 16 at r2](https://reviewable.io:443/reviews/robotlocomotion/drake/4405#-KYPZtqN704IwlB4lqn8:-KYQ2hFr4Saotf26CZ5Y:b-4znkk0) ([raw file](https://github.com/robotlocomotion/drake/blob/d83d147a017e12987f450795da93abbd9977b372/drake/examples/kuka_iiwa_arm/iiwa_lcm.h#L16)):*
<details><summary><i>Previously, sammy-tri (Sam Creasey) wrote…</i></summary><blockquote>

That would be my assumption, since it's also what I think we'd do if we had multiple actual iiwa arms...   though given that LCM is typically multicast, running multiple iiwa arms would take some additional code changes or careful separation of the LCM domains (e.g. different UDP port numbers).
</blockquote></details>

Good point! We'll need to modify our LCM Mocking framework to allow us to specify the multicast address! #4417 

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/robotlocomotion/drake/4405)*
<!-- Sent from Reviewable.io -->

liangfok commented:
Reviewed 1 of 3 files at r3.
Review status: all files reviewed at latest revision, all discussions resolved.

---



*Comments from [Reviewable](https://reviewable.io:443/reviews/robotlocomotion/drake/4405)*
<!-- Sent from Reviewable.io -->

sammy-tri commented:
Review status: all files reviewed at latest revision, all discussions resolved.

---

*[drake/examples/kuka_iiwa_arm/iiwa_lcm.h, line 16 at r2](https://reviewable.io:443/reviews/robotlocomotion/drake/4405#-KYPZtqN704IwlB4lqn8:-KYQ5EHi1m4PulJKv7qm:b8xykr8) ([raw file](https://github.com/robotlocomotion/drake/blob/d83d147a017e12987f450795da93abbd9977b372/drake/examples/kuka_iiwa_arm/iiwa_lcm.h#L16)):*
<details><summary><i>Previously, liangfok (Chien-Liang Fok) wrote…</i></summary><blockquote>

Good point! We'll need to modify our LCM Mocking framework to allow us to specify the multicast address! #4417 
</blockquote></details>

There are other ways (such as using different channel names or having a robot specifier in the message) too.  I haven't given it a ton of thought yet, though it is coming...

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/robotlocomotion/drake/4405)*
<!-- Sent from Reviewable.io -->

liangfok commented:
Review status: all files reviewed at latest revision, 1 unresolved discussion.

---

*[drake/examples/kuka_iiwa_arm/iiwa_lcm.h, line 16 at r2](https://reviewable.io:443/reviews/robotlocomotion/drake/4405#-KYPZtqN704IwlB4lqn8:-KYQ5iMtDEXxFeFrBpvi:b9h503v) ([raw file](https://github.com/robotlocomotion/drake/blob/d83d147a017e12987f450795da93abbd9977b372/drake/examples/kuka_iiwa_arm/iiwa_lcm.h#L16)):*
<details><summary><i>Previously, sammy-tri (Sam Creasey) wrote…</i></summary><blockquote>

There are other ways (such as using different channel names or having a robot specifier in the message) too.  I haven't given it a ton of thought yet, though it is coming...
</blockquote></details>

I look forward to hearing your thoughts!

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/robotlocomotion/drake/4405)*
<!-- Sent from Reviewable.io -->

jwnimmer-tri commented:
*[drake/examples/kuka_iiwa_arm/iiwa_lcm.cc, line 15 at r2](https://reviewable.io:443/reviews/robotlocomotion/drake/4405#-KYPg5F_6-0JGttk3UcR:-KYQ8yJQ2VjAKwJdm6fZ:bmncshi) ([raw file](https://github.com/robotlocomotion/drake/blob/d83d147a017e12987f450795da93abbd9977b372/drake/examples/kuka_iiwa_arm/iiwa_lcm.cc#L15)):*
<details><summary><i>Previously, liangfok (Chien-Liang Fok) wrote…</i></summary><blockquote>

OK. Let's keep this in mind as a possible future enhancement. I feel our current usage of port number makes potentially brittle assumptions about how `systems::System` implements ports. As it stands, `systems::Systems::DeclareOutputPort()` does not promise that the first declared port will be at index 0!
</blockquote></details>

I agree with Sam.  I think assuming zero is correct for user code.  A different PR can fix the System API docs if we feel we need to document the "0,1,...N" counting.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/robotlocomotion/drake/4405)*
<!-- Sent from Reviewable.io -->

liangfok commented:
Review status: all files reviewed at latest revision, 1 unresolved discussion.

---

*[drake/examples/kuka_iiwa_arm/iiwa_lcm.cc, line 15 at r2](https://reviewable.io:443/reviews/robotlocomotion/drake/4405#-KYPg5F_6-0JGttk3UcR:-KYQ9atv5A_z1gYCFuAb:b-3ldcie) ([raw file](https://github.com/robotlocomotion/drake/blob/d83d147a017e12987f450795da93abbd9977b372/drake/examples/kuka_iiwa_arm/iiwa_lcm.cc#L15)):*
<details><summary><i>Previously, jwnimmer-tri (Jeremy Nimmer) wrote…</i></summary><blockquote>

I agree with Sam.  I think assuming zero is correct for user code.  A different PR can fix the System API docs if we feel we need to document the "0,1,...N" counting.
</blockquote></details>

OK I'll submit a PR soon updating the API's contract with users.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/robotlocomotion/drake/4405)*
<!-- Sent from Reviewable.io -->

sammy-tri commented:
Review status: all files reviewed at latest revision, 1 unresolved discussion.

---

*[a discussion](https://reviewable.io:443/reviews/robotlocomotion/drake/4405#-KYPhoV-2VD62-chCJfU:-KYQIoBN2av4FeE89B-B:bqvuaa4) (no related file):*
<details><summary><i>Previously, liangfok (Chien-Liang Fok) wrote…</i></summary><blockquote>

OK thanks.
</blockquote></details>

Ugh.  Either I'm missing something obvious or this is going to be a *really* non-trivial unit test to write.  I'm having trouble even finding an example of a unit test of a System where the value of an input port varies (without creating a Diagram).  I sincerely hope I'm missing an obvious example, or we're going to have real trouble keeping even poor quality unit tests around if "change my input value, simulate for some time, check my new output value" proves to be a challenging test to write.

I really hope I'm missing something.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/robotlocomotion/drake/4405)*
<!-- Sent from Reviewable.io -->

david-german-tri commented:
Review status: all files reviewed at latest revision, 1 unresolved discussion.

---

*[a discussion](https://reviewable.io:443/reviews/robotlocomotion/drake/4405#-KYPhoV-2VD62-chCJfU:-KYQLkUJDpHzYYEHOHlP:bwf2fpp) (no related file):*
<details><summary><i>Previously, sammy-tri (Sam Creasey) wrote…</i></summary><blockquote>

Ugh.  Either I'm missing something obvious or this is going to be a *really* non-trivial unit test to write.  I'm having trouble even finding an example of a unit test of a System where the value of an input port varies (without creating a Diagram).  I sincerely hope I'm missing an obvious example, or we're going to have real trouble keeping even poor quality unit tests around if "change my input value, simulate for some time, check my new output value" proves to be a challenging test to write.

I really hope I'm missing something.
</blockquote></details>

Does FixInputPort help?
  https://github.com/RobotLocomotion/drake/blob/master/drake/systems/framework/context.h#L185

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/robotlocomotion/drake/4405)*
<!-- Sent from Reviewable.io -->

liangfok commented:
Review status: all files reviewed at latest revision, 1 unresolved discussion.

---

*[a discussion](https://reviewable.io:443/reviews/robotlocomotion/drake/4405#-KYPhoV-2VD62-chCJfU:-KYQbs35-6ih_W2RbAsZ:b-haclc) (no related file):*
<details><summary><i>Previously, david-german-tri (David German) wrote…</i></summary><blockquote>

Does FixInputPort help?
  https://github.com/RobotLocomotion/drake/blob/master/drake/systems/framework/context.h#L185
</blockquote></details>

Also, could this serve as an example?

https://github.com/RobotLocomotion/drake/blob/4f64ccc62a662b6b32a4fe6e7663b1d8e78a35f5/drake/multibody/rigid_body_plant/test/rigid_body_plant_test.cc#L264

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/robotlocomotion/drake/4405)*
<!-- Sent from Reviewable.io -->

liangfok commented:
Review status: all files reviewed at latest revision, 2 unresolved discussions.

---

*[drake/examples/kuka_iiwa_arm/iiwa_lcm.cc, line 15 at r2](https://reviewable.io:443/reviews/robotlocomotion/drake/4405#-KYPg5F_6-0JGttk3UcR:-KYRW39_ZgtpnLGY4akr:b-9kuk15) ([raw file](https://github.com/robotlocomotion/drake/blob/d83d147a017e12987f450795da93abbd9977b372/drake/examples/kuka_iiwa_arm/iiwa_lcm.cc#L15)):*
<details><summary><i>Previously, liangfok (Chien-Liang Fok) wrote…</i></summary><blockquote>

OK I'll submit a PR soon updating the API's contract with users.
</blockquote></details>

Done: #4423.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/robotlocomotion/drake/4405)*
<!-- Sent from Reviewable.io -->

sammy-tri commented:
Review status: all files reviewed at latest revision, 1 unresolved discussion.

---

*[a discussion](https://reviewable.io:443/reviews/robotlocomotion/drake/4405#-KYPhoV-2VD62-chCJfU:-KYTqF5g0LTdJLnNxvPf:bvp3g84) (no related file):*
<details><summary><i>Previously, liangfok (Chien-Liang Fok) wrote…</i></summary><blockquote>

Also, could this serve as an example?

https://github.com/RobotLocomotion/drake/blob/4f64ccc62a662b6b32a4fe6e7663b1d8e78a35f5/drake/multibody/rigid_body_plant/test/rigid_body_plant_test.cc#L264
</blockquote></details>

Hmm...  It hadn't occurred to me that I could call FixInputPort repeatedly to change the value, since SetInputPort is documented to replace the port with the new value...  I guess I can write a helper function which takes the new values (for both the abstract and the vector valued input ports), creates a new FreestandingInputPort for the abstract one, and FixInputPort's both of them to the new value before updating the simulation (or calling EvalDifferenceUpdates and EvalOutput manually (EvalOutput won't magically cause EvalDifferenceUpdates to run, I assume?)).  That's a lot of messing around with creating and destroying ports, but it's not so bad if I bury it out of sight I guess.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/robotlocomotion/drake/4405)*
<!-- Sent from Reviewable.io -->

ggould-tri commented:
Review status: all files reviewed at latest revision, 1 unresolved discussion.

---

*[a discussion](https://reviewable.io:443/reviews/robotlocomotion/drake/4405#-KYPhoV-2VD62-chCJfU:-KYTwZ4E8AcX9sW5-iu1:b-snaqj6) (no related file):*
<details><summary><i>Previously, sammy-tri (Sam Creasey) wrote…</i></summary><blockquote>

Hmm...  It hadn't occurred to me that I could call FixInputPort repeatedly to change the value, since SetInputPort is documented to replace the port with the new value...  I guess I can write a helper function which takes the new values (for both the abstract and the vector valued input ports), creates a new FreestandingInputPort for the abstract one, and FixInputPort's both of them to the new value before updating the simulation (or calling EvalDifferenceUpdates and EvalOutput manually (EvalOutput won't magically cause EvalDifferenceUpdates to run, I assume?)).  That's a lot of messing around with creating and destroying ports, but it's not so bad if I bury it out of sight I guess.
</blockquote></details>

If it is super difficult to create decent unit tests for diagrams, that should be filed as an issue.  Once that issue is filed, much as I might find it distasteful I don't think a lack of tests is fatal to code in `examples` which is pretty much full of undertested stuff.

A compromise I used in the Schunk code was to just test that under constant input the system eventually reached and stayed at its expected endpoint.  That's not by any means comprehensive but it was enough to smoke test and sanity check things.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/robotlocomotion/drake/4405)*
<!-- Sent from Reviewable.io -->

ggould-tri commented:
Reviewed 1 of 3 files at r3.
Review status: all files reviewed at latest revision, 5 unresolved discussions.

---

*[drake/examples/schunk_wsg/schunk_wsg_lcm.cc, line 107 at r3](https://reviewable.io:443/reviews/robotlocomotion/drake/4405#-KYU-v5fChg_bx9ifQ_U:-KYU-v5fCFO2Wituzk3h:befm0ei) ([raw file](https://github.com/robotlocomotion/drake/blob/1d155374153fe25d701d54104af09916e71fabd4/drake/examples/schunk_wsg/schunk_wsg_lcm.cc#L107)):*
> ```cpp
>   const double kMaxVelocity = 0.42;  // m/s
>   const double kMaxAccel = 5.;  // m/s^2
>   const double kTimeToMaxAccel = kMaxVelocity / kMaxAccel;
> ```

I am confused by this name.  It looks to me like the time, at max accel, to reach max velocity.  Can you clarify, document, or rename?

---

*[drake/examples/schunk_wsg/schunk_wsg_lcm.cc, line 110 at r3](https://reviewable.io:443/reviews/robotlocomotion/drake/4405#-KYU0JIj-qAR0talEhOx:-KYU0JIjCWBs88lmZZpH:b63g9hs) ([raw file](https://github.com/robotlocomotion/drake/blob/1d155374153fe25d701d54104af09916e71fabd4/drake/examples/schunk_wsg/schunk_wsg_lcm.cc#L110)):*
> ```cpp
>   // TODO(sam.creasey) this should probably consider current speed
>   // if the gripper is already moving.
>   const double kDistanceToMaxAccel =
> ```

Same as above -- as I read this formula, it is distance traveled to reach max V, not max A.  ??

---

*[drake/examples/schunk_wsg/schunk_wsg_lcm.cc, line 128 at r3](https://reviewable.io:443/reviews/robotlocomotion/drake/4405#-KYTzeG1AN0m-sT6mqtY:-KYTzeG1EypNgMg5RTgT:bunor5u) ([raw file](https://github.com/robotlocomotion/drake/blob/1d155374153fe25d701d54104af09916e71fabd4/drake/examples/schunk_wsg/schunk_wsg_lcm.cc#L128)):*
> ```cpp
>   if (delta < kDistanceToMaxAccel * 2) {
>     // If we can't accelerate to our maximum (and decelerate again)
>     // within the target travel distance, calculate a the peak
> ```

fyi - typo 'a the'

---

*[drake/examples/schunk_wsg/schunk_wsg_lcm.cc, line 192 at r3](https://reviewable.io:443/reviews/robotlocomotion/drake/4405#-KYU-2XM6uN2DnLhcHiT:-KYU-2XM4EQicYAgWLgG:bv8mjaf) ([raw file](https://github.com/robotlocomotion/drake/blob/1d155374153fe25d701d54104af09916e71fabd4/drake/examples/schunk_wsg/schunk_wsg_lcm.cc#L192)):*
> ```cpp
>   status.actual_position_mm = -2 * state->GetAtIndex(position_index_) * 1e3;
>   // TODO(sam.creasey) Figure out how to get the actual force from
>   // the plant so that we can populate this field.
> ```

FYI -- if the actual force is not available from the RBP, you should file an issue for that soonish, since RBP redesign is in progress.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/robotlocomotion/drake/4405)*
<!-- Sent from Reviewable.io -->

sammy-tri commented:
Review status: 4 of 10 files reviewed at latest revision, 4 unresolved discussions.

---

*[a discussion](https://reviewable.io:443/reviews/robotlocomotion/drake/4405#-KYPhoV-2VD62-chCJfU:-KYUjn4i8IN_bkSXHU_O:b-f6fkao) (no related file):*
<details><summary><i>Previously, ggould-tri wrote…</i></summary><blockquote>

If it is super difficult to create decent unit tests for diagrams, that should be filed as an issue.  Once that issue is filed, much as I might find it distasteful I don't think a lack of tests is fatal to code in `examples` which is pretty much full of undertested stuff.

A compromise I used in the Schunk code was to just test that under constant input the system eventually reached and stayed at its expected endpoint.  That's not by any means comprehensive but it was enough to smoke test and sanity check things.
</blockquote></details>

Added a simple test using constant input.

---

*[drake/examples/kuka_iiwa_arm/iiwa_lcm.h, line 17 at r1](https://reviewable.io:443/reviews/robotlocomotion/drake/4405#-KYP5kQS40yS9yU_6hkV:-KYUnVSYDH4ValADoIOa:b-896fix) ([raw file](https://github.com/robotlocomotion/drake/blob/7b3e84d1a61022dd2808fdc7a16847d0258d32cb/drake/examples/kuka_iiwa_arm/iiwa_lcm.h#L17)):*
<details><summary><i>Previously, liangfok (Chien-Liang Fok) wrote…</i></summary><blockquote>

Since we can't think of a reason the number of joints would change, I recommend removing this parameter and using a `kNumJoints` constant since that will eliminate one source of user error.
</blockquote></details>

Done.

---

*[drake/examples/schunk_wsg/schunk_wsg_lcm.cc, line 107 at r3](https://reviewable.io:443/reviews/robotlocomotion/drake/4405#-KYU-v5fChg_bx9ifQ_U:-KYUkj9x13OUH7GUPvD9:b-896fix) ([raw file](https://github.com/robotlocomotion/drake/blob/1d155374153fe25d701d54104af09916e71fabd4/drake/examples/schunk_wsg/schunk_wsg_lcm.cc#L107)):*
<details><summary><i>Previously, ggould-tri wrote…</i></summary><blockquote>

I am confused by this name.  It looks to me like the time, at max accel, to reach max velocity.  Can you clarify, document, or rename?
</blockquote></details>

Done.

---

*[drake/examples/schunk_wsg/schunk_wsg_lcm.cc, line 110 at r3](https://reviewable.io:443/reviews/robotlocomotion/drake/4405#-KYU0JIj-qAR0talEhOx:-KYUkjYP9uuvjDiPjPF9:b-896fix) ([raw file](https://github.com/robotlocomotion/drake/blob/1d155374153fe25d701d54104af09916e71fabd4/drake/examples/schunk_wsg/schunk_wsg_lcm.cc#L110)):*
<details><summary><i>Previously, ggould-tri wrote…</i></summary><blockquote>

Same as above -- as I read this formula, it is distance traveled to reach max V, not max A.  ??
</blockquote></details>

Done.

---

*[drake/examples/schunk_wsg/schunk_wsg_lcm.cc, line 128 at r3](https://reviewable.io:443/reviews/robotlocomotion/drake/4405#-KYTzeG1AN0m-sT6mqtY:-KYUkn687BzdVuVH2sMZ:b-896fix) ([raw file](https://github.com/robotlocomotion/drake/blob/1d155374153fe25d701d54104af09916e71fabd4/drake/examples/schunk_wsg/schunk_wsg_lcm.cc#L128)):*
<details><summary><i>Previously, ggould-tri wrote…</i></summary><blockquote>

fyi - typo 'a the'
</blockquote></details>

Done.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/robotlocomotion/drake/4405)*
<!-- Sent from Reviewable.io -->

jamiesnape commented:
@drake-jenkins-bot retest this please

liangfok commented:
<img class="emoji" title=":lgtm:" alt=":lgtm:" align="absmiddle" src="https://reviewable.io/lgtm.png" height="20" width="61"/> I let a few BTW comments.

---

Reviewed 5 of 6 files at r4.
Review status: 9 of 10 files reviewed at latest revision, 10 unresolved discussions.

---

*[a discussion](https://reviewable.io:443/reviews/robotlocomotion/drake/4405#-KYPhoV-2VD62-chCJfU:-KYV0IQNSXJWmHE4egJq:b-ucfknh) (no related file):*
<details><summary><i>Previously, sammy-tri (Sam Creasey) wrote…</i></summary><blockquote>

Added a simple test using constant input.
</blockquote></details>

Thanks!

---

*[drake/examples/kuka_iiwa_arm/iiwa_lcm.h, line 15 at r4](https://reviewable.io:443/reviews/robotlocomotion/drake/4405#-KYUy74bGmTEbhHtiLuH:-KYUy74bGmTEbhHtiLuI:b-ovlfcc) ([raw file](https://github.com/robotlocomotion/drake/blob/b7129271b00bbe3b654862bf33d51e58e6ad6b4d/drake/examples/kuka_iiwa_arm/iiwa_lcm.h#L15)):*
BTW: Consider replacing:

> LCM message has arrived with a command

with:

> lcmt_iiwa_command message was received

---

*[drake/examples/schunk_wsg/schunk_wsg_lcm.cc, line 105 at r4](https://reviewable.io:443/reviews/robotlocomotion/drake/4405#-KYUxSRZoG65jsI9W2D4:-KYUxSRZoG65jsI9W2D5:bdb1m4) ([raw file](https://github.com/robotlocomotion/drake/blob/b7129271b00bbe3b654862bf33d51e58e6ad6b4d/drake/examples/schunk_wsg/schunk_wsg_lcm.cc#L105)):*
> ```cpp
>   // values available for manual control through the gripper's web
>   // interface.
>   const double kMaxVelocity = 0.42;  // m/s
> ```

BTW: nit: given that this variable's name contains the word "Velocity", does the `// m/s` comment add anything?

(same comment below regarding `// m/s^2`)

---

*[drake/examples/schunk_wsg/schunk_wsg_lcm.h, line 18 at r4](https://reviewable.io:443/reviews/robotlocomotion/drake/4405#-KYUzrG5jA3aa2lTYH7c:-KYUzrG5jA3aa2lTYH7d:b-dn2kea) ([raw file](https://github.com/robotlocomotion/drake/blob/b7129271b00bbe3b654862bf33d51e58e6ad6b4d/drake/examples/schunk_wsg/schunk_wsg_lcm.h#L18)):*
> ```objc
> // state of the gripper (particularly the maximum force).
> 
> /// Receives commands for a Schunk WSG (input port 0) along with the
> ```

BTW, consider adding `lcmt_schunk_wsg_command` in front of "commands" so the reader knows that input port 0 is (1) an abstract input port and (2) the type of data received through the input port.

---

*[drake/examples/schunk_wsg/test/schunk_wsg_lcm_test.cc, line 26 at r4](https://reviewable.io:443/reviews/robotlocomotion/drake/4405#-KYUyy16Xy6vyh-DOQ_h:-KYUyy16Xy6vyh-DOQ_i:bh42kpe) ([raw file](https://github.com/robotlocomotion/drake/blob/b7129271b00bbe3b654862bf33d51e58e6ad6b4d/drake/examples/schunk_wsg/test/schunk_wsg_lcm_test.cc#L26)):*
> ```cpp
>   // Start off with the gripper closed (zero) and a command to open to
>   // 100mm.
>   lcmt_schunk_wsg_command initial_command{};
> ```

BTW, since `lcmt_schunk_wsg_command` is not POD, the `{}` is not necessary.

---

*[drake/examples/schunk_wsg/test/schunk_wsg_lcm_test.cc, line 36 at r4](https://reviewable.io:443/reviews/robotlocomotion/drake/4405#-KYV-mLVtJnzslQiSfZY:-KYV-mLVtJnzslQiSfZZ:bv8z0un) ([raw file](https://github.com/robotlocomotion/drake/blob/b7129271b00bbe3b654862bf33d51e58e6ad6b4d/drake/examples/schunk_wsg/test/schunk_wsg_lcm_test.cc#L36)):*
> ```cpp
>   context->FixInputPort(1, Eigen::VectorXd::Zero(1));
> 
>   // Step a little bit. We should be commanding a point on the
> ```

BTW, consider replacing "Step" with "Simulate forward in time".

---

*[drake/examples/schunk_wsg/test/schunk_wsg_lcm_test.cc, line 38 at r4](https://reviewable.io:443/reviews/robotlocomotion/drake/4405#-KYV-YYk6XiaLTObY5Rq:-KYV-YYlZC9rn8HUoqRf:b2sv3ba) ([raw file](https://github.com/robotlocomotion/drake/blob/b7129271b00bbe3b654862bf33d51e58e6ad6b4d/drake/examples/schunk_wsg/test/schunk_wsg_lcm_test.cc#L38)):*
> ```cpp
>   // Step a little bit. We should be commanding a point on the
>   // trajectory wider than zero, but not to the target yet.
>   const double expected_target = -0.05;  // 50mm
> ```

BTW, consider adding a comment explaining why the expected target is negative while the actual target is positive (100mm).

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/robotlocomotion/drake/4405#-:-KYV0G8G5PRS-5EJrKoG:bhf1g6l)*
<!-- Sent from Reviewable.io -->

liangfok commented:
Reviewed 1 of 6 files at r4.
Review status: all files reviewed at latest revision, 10 unresolved discussions.

---



*Comments from [Reviewable](https://reviewable.io:443/reviews/robotlocomotion/drake/4405)*
<!-- Sent from Reviewable.io -->

sammy-tri commented:
Review status: 9 of 10 files reviewed at latest revision, 8 unresolved discussions.

---

*[drake/examples/kuka_iiwa_arm/iiwa_lcm.h, line 15 at r4](https://reviewable.io:443/reviews/robotlocomotion/drake/4405#-KYUy74bGmTEbhHtiLuH:-KYV3tII7byl9wHWdvu3:b-mcjekq) ([raw file](https://github.com/robotlocomotion/drake/blob/b7129271b00bbe3b654862bf33d51e58e6ad6b4d/drake/examples/kuka_iiwa_arm/iiwa_lcm.h#L15)):*
<details><summary><i>Previously, liangfok (Chien-Liang Fok) wrote…</i></summary><blockquote>

BTW: Consider replacing:

> LCM message has arrived with a command

with:

> lcmt_iiwa_command message was received

</blockquote></details>

The message type is explicitly stated above...  I think it's probably clear.

---

*[drake/examples/schunk_wsg/schunk_wsg_lcm.cc, line 105 at r4](https://reviewable.io:443/reviews/robotlocomotion/drake/4405#-KYUxSRZoG65jsI9W2D4:-KYV37By5gffGKe4YumL:b-37pufm) ([raw file](https://github.com/robotlocomotion/drake/blob/b7129271b00bbe3b654862bf33d51e58e6ad6b4d/drake/examples/schunk_wsg/schunk_wsg_lcm.cc#L105)):*
<details><summary><i>Previously, liangfok (Chien-Liang Fok) wrote…</i></summary><blockquote>

BTW: nit: given that this variable's name contains the word "Velocity", does the `// m/s` comment add anything?

(same comment below regarding `// m/s^2`)
</blockquote></details>

probably not, but I don't see how noting what I intended for units would be harmful...

---

*[drake/examples/schunk_wsg/schunk_wsg_lcm.h, line 18 at r4](https://reviewable.io:443/reviews/robotlocomotion/drake/4405#-KYUzrG5jA3aa2lTYH7c:-KYV3Lws3VTEBg7fObYu:b-896fix) ([raw file](https://github.com/robotlocomotion/drake/blob/b7129271b00bbe3b654862bf33d51e58e6ad6b4d/drake/examples/schunk_wsg/schunk_wsg_lcm.h#L18)):*
<details><summary><i>Previously, liangfok (Chien-Liang Fok) wrote…</i></summary><blockquote>

BTW, consider adding `lcmt_schunk_wsg_command` in front of "commands" so the reader knows that input port 0 is (1) an abstract input port and (2) the type of data received through the input port.
</blockquote></details>

Done.

---

*[drake/examples/schunk_wsg/test/schunk_wsg_lcm_test.cc, line 26 at r4](https://reviewable.io:443/reviews/robotlocomotion/drake/4405#-KYUyy16Xy6vyh-DOQ_h:-KYV3N5A6iUGWmghqUK8:bcx3lql) ([raw file](https://github.com/robotlocomotion/drake/blob/b7129271b00bbe3b654862bf33d51e58e6ad6b4d/drake/examples/schunk_wsg/test/schunk_wsg_lcm_test.cc#L26)):*
<details><summary><i>Previously, liangfok (Chien-Liang Fok) wrote…</i></summary><blockquote>

BTW, since `lcmt_schunk_wsg_command` is not POD, the `{}` is not necessary.
</blockquote></details>

It should still cause the members of the class to be initialized to zero though, right?  The LCM struct compiler doesn't emit a constructor to do the initialization AFAICT.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/robotlocomotion/drake/4405)*
<!-- Sent from Reviewable.io -->

ggould-tri commented:
<img class="emoji" title=":lgtm:" alt=":lgtm:" align="absmiddle" src="https://reviewable.io/lgtm.png" height="20" width="61"/>

---

Review status: 9 of 10 files reviewed at latest revision, 3 unresolved discussions.

---



*Comments from [Reviewable](https://reviewable.io:443/reviews/robotlocomotion/drake/4405#-:-KYV6dOTBLfMgR0v40A6:bnfp4nl)*
<!-- Sent from Reviewable.io -->

liangfok commented:
Reviewed 1 of 1 files at r5.
Review status: all files reviewed at latest revision, 2 unresolved discussions.

---

*[drake/examples/kuka_iiwa_arm/iiwa_lcm.h, line 15 at r4](https://reviewable.io:443/reviews/robotlocomotion/drake/4405#-KYUy74bGmTEbhHtiLuH:-KYVBrppVcitgFXonl5p:br052pj) ([raw file](https://github.com/robotlocomotion/drake/blob/b7129271b00bbe3b654862bf33d51e58e6ad6b4d/drake/examples/kuka_iiwa_arm/iiwa_lcm.h#L15)):*
<details><summary><i>Previously, sammy-tri (Sam Creasey) wrote…</i></summary><blockquote>

The message type is explicitly stated above...  I think it's probably clear.
</blockquote></details>

I suppose the risk is for the reader to be left confused about whether it's possible for an LCM message to arrive that does not contain a command. I'll leave this to your discretion.

---

*[drake/examples/schunk_wsg/test/schunk_wsg_lcm_test.cc, line 26 at r4](https://reviewable.io:443/reviews/robotlocomotion/drake/4405#-KYUyy16Xy6vyh-DOQ_h:-KYVB_82XlHyof3Wcjrl:b7dlp6x) ([raw file](https://github.com/robotlocomotion/drake/blob/b7129271b00bbe3b654862bf33d51e58e6ad6b4d/drake/examples/schunk_wsg/test/schunk_wsg_lcm_test.cc#L26)):*
<details><summary><i>Previously, sammy-tri (Sam Creasey) wrote…</i></summary><blockquote>

It should still cause the members of the class to be initialized to zero though, right?  The LCM struct compiler doesn't emit a constructor to do the initialization AFAICT.
</blockquote></details>

You're right that the LCM struct compiler doesn't emit code that does the initialization. Sounds like something we should fix upstream! I'll leave this to your discretion.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/robotlocomotion/drake/4405)*
<!-- Sent from Reviewable.io -->

ggould-tri commented:
Reviewed 1 of 1 files at r5.
Review status: all files reviewed at latest revision, 2 unresolved discussions.

---



*Comments from [Reviewable](https://reviewable.io:443/reviews/robotlocomotion/drake/4405)*
<!-- Sent from Reviewable.io -->
